### PR TITLE
fix: task limit

### LINKED ISSUE #28: Properly handle task limit
The original implementation allowed for productive contributors to work on two tasks concurrently. With a starting two task limit:
| Condition                                              | Task Adjustment       | Explanation                                                                                             |
|--------------------------------------------------------|-----------------------|---------------------------------------------------------------------------------------------------------|
| If any reviewer approved the pull                      | +1 task per pull      | It's likely that the pull is good, just waiting on other slow reviewers to confirm.                      |
| If reviewers take longer than 24 hours                 | +1 task per pull      | Don't wait for lazy reviewers to be able to start a new task.                                            |
| If any reviewer requested changes                      | -1 task per pull      | They should focus on addressing the changes as a top priority.                                           |

_Originally posted by @0x4007 in https://github.com/ubiquibot/command-start-stop/pull/19#discussion_r1731441502_
            

For the sake of this conversation a completed PR is when the conditions are met in the above table.

In the current state the code checks if any review is approved or if there is 0 reviews but 24 hours have passed since the creation of the PR, however it doesn't check if there's any requested changes so it will count it as completed even if there 1 approve and 1 requested changes.

There is still a problem when the reviewer requests changes and the contributor resolves those changes but waits for the reviewer to make a new review so they can't start another task.
There's two possible solutions: check if the reviewer was requested more than x hours ago or check that changes have been marked as solved

### PR DESCRIPTION
**Pull request by gentlementlegen**
Resolves #28
Depends on #86 

QA: https://github.com/Meniole/command-start-stop/issues/48

## What are the changes
- if a PR has a change requested, counts as -1
- if a PR was opened with no reviews, either -1 is less than the review threshold or +1 if passed
- if a PR has approved status (without any other required changes), +1
- if a PR was reviewed, changed, then review requested again, either -1 is less than the review threshold or +1 if passed

### CODE REVIEW

**Comment by gentlementlegen** [comment]
@whilefoo Could you please also have a look?

**Comment by 0x4007** [comment]
Your QA isn't clear can you list the three scenarios and their outputs?

File: src/types/plugin-input.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ function maxConcurrentTasks() {
 
 export const pluginSettingsSchema = T.Object(
   {
-    reviewDelayTolerance: T.String({ default: "1 Day" }),
+    reviewDelayTolerance: T.String({ default: "1 Day", description: "How long shall the wait be for a reviewer to take action?" }),
```
Comment (0x4007) [inline src/types/plugin-input.ts:42]: This description isn't very clear. Given the specification, I guess that this should describe the max time we will wait for a reviewer to review while holding up the task for the assignee

**Review by 0x4007 — COMMENTED** [review]
Code looks fine I just wish the QA was more clear.

**Comment by gentlementlegen** [comment]
@0x4007 It is hard to do a clear QA because it depends on a pull-request review. So I had to ask for review, test after idle on a review, or after a comment. The linked review is https://github.com/Meniole/command-start-stop/pull/42

**Comment by 0x4007** [comment]
Then let's test in prod

File: src/types/plugin-input.ts (hunk 1 above)
Comment (gentlementlegen) [reply src/types/plugin-input.ts:42]: Maybe something like "Duration delay for a pull-request to count against the user's maximum assigned issues quota" ?

**Pull request merged**